tests/test_E4_verification.py

```diff
@@ -35,3 +37,71 @@ def test_minimal_renewable_share_test_fails():
     }
     return_value = E4.minimal_renewable_share_test(dict_values)
     assert return_value == False
+
+
+def test_detect_excessive_excess_generation_in_bus_warning_is_logged(caplog):
+    """A logging.warning is printed due to excess generation. """
+    bus_label = "Test_bus"
+    dict_values = {
+        "optimizedFlows": {
+            bus_label: pd.DataFrame(
+                {"inflow": [1, 2, 3], "outflow": [-1, -1, -2], "excess": [0, 1, 1]}
+            )
+        }
+    }
+    with caplog.at_level(logging.WARNING):
+        E4.detect_excessive_excess_generation_in_bus(dict_values=dict_values)
+    assert (
+        f"Attention, on bus {bus_label} there is an excessive excess generation"
+        in caplog.text
+    ), f"An intended warning is not logged although there is excess generation."
+
+
+def test_detect_excessive_excess_generation_in_bus_no_excess(caplog):
+    """No excess generation takes place. """
+    bus_label = "Test_bus"
+    dict_values = {
+        "optimizedFlows": {
+            bus_label: pd.DataFrame(
+                {
+                    "inflow": [1, 2, 3],
+                    "outflow": [-1, -1.9, -2.5],
+                    "excess": [0, 0.1, 0.5],
+                }
+            )
+        }
+    }
+    with caplog.at_level(logging.WARNING):
+        E4.detect_excessive_excess_generation_in_bus(dict_values=dict_values)
+    assert (
+        caplog.text == ""
+    ), f"A warning is logged although there is no excess generation."
+
+
+def test_detect_excessive_excess_generation_in_bus_several_busses_two_warnings(caplog):
+    """Excess generation takes place in two busses. """
+    dict_values = {
+        "optimizedFlows": {
+            "Bus_without_excess": pd.DataFrame(
+                {
+                    "inflow": [1, 2, 3],
+                    "outflow": [-1, -1.9, -2.5],
+                    "excess": [0, 0.1, 0.5],
+                }
+            ),
+            "Bus_with_excess_1": pd.DataFrame(
+                {"inflow": [1, 2, 3], "outflow": [-1, -1, -2], "excess": [0, 1, 1],}
+            ),
+            "Bus_with_excess_2": pd.DataFrame(
+                {"inflow": [1, 2, 3], "outflow": [-1, -1, -1], "excess": [0, 1, 2],}
+            ),
+        }
+    }
+    with caplog.at_level(logging.WARNING):
+        E4.detect_excessive_excess_generation_in_bus(dict_values=dict_values)
+    assert (
+        f"Attention, on bus Bus_with_excess_1 there is an excessive excess generation"
+        in caplog.text
+        and f"Attention, on bus Bus_with_excess_2 there is an excessive excess generation"
+        in caplog.text
```

Comment:
I would define a string `Bus_with_excess_1="Bus_with_excess_1") so that you can use it for the json definition as well as for the warning message check.

Reply:
Same for Bus_with_excess_2.